### ISSUE TITLE
Enhancement request: Allow random generation of MAC Address

### ISSUE THREAD

**Issue by jguillaumes** [simh/simh#317]
Hello,

This is not a bug report, but an ehnancement idea.

Allow random generation of the MAC address of the simulated NIC interfaces (XU, XQ, and so on). 

Rationale: 

If an user runs multiple simulators under the same network (PCAP or VDE specially) the user has to be careful to manually edit each configuration file to allocate each simulator an unique MAC address.

That prevents, for example, automatic provisioning of simulated machines. ie, if someone wants to publish "public" images for simh, the final user has to review manually the MAC addresses,

I'd say this is a LOW priority request, and after all it is basically an estetic enhancement, so feel free to ignore/discard/reject it.

**Comment by markpizz** [comment]
I've thought about this before.

There are several complications:

1. The MAC address on the device is, in general for the systems being simulated by simh, just a briefly used value which is subsequently dynamically changed by some component of the network stack to leverage the DECnet address of the system.  Dynamic assignment of the DECnet address then becomes essentially the same problem.
2. Ignoring point 1, if this was to work, it would be very nice if the dynamically assigned address were actually the same value from one startup of a particular simulator instance to the next.
3. The concept of MAC address is that it actually should be a globally unique number that can reliably be counted on for its uniqueness.  It would seem that it is better to be never unique (as it is now) instead of possibly unique.

Suggest how you think it should work given these points.

**Comment by jguillaumes** [comment]
Good points indeed.

1) That's true for the systems which support DECNET. For UNIX like systems (ancient UNIX basically) and modern BSDs this point does not apply. Nevertheless, the option to assign an user specified MAC address should also exist.

2) Actually, this is less important as it seems. Some modern hardware (for instance, most ARM based small computers) actually assign a random MAC to their ethernet interfaces. But, we are talking about old stuff, so, yes it shold be preserved. 

How? Probably the easiest thing that comes to my mind is to "attach" the randomly generated XQ or XU to a file where the generated MAC would be stored. Something like

SET XQ MAC=random[:filespec]

If filespec exists and contains a valid generated MAC, simh would use that one. If filespec does not exist or its content is not valid, it would generate a new MAC, store it into filespec and issue an informative message to console.

filespec could (or not) be optional. If not present, a random MAC would be always generated.

3) Hmmm this one is the hardest. Some ideas come to mind, Probably not perfect but...

a) Use a combination of the host MAC address and the full path of the configuration file to compute a HASH and use that HASH as seed for the generating algorithm.

Problem: it is not immediate what is the "host MAC address". Specially if you consider the host machine does not need have a NIC... (but... which machine does not have one nowadays?).

b) Instead of the host MAC address, use a unique identifier for the machine or, barred that, for the system disk:

- In OSX, use gethostuuid()
- In windows, use HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\MachineGuid
- In Linux the best chance would be probably to use libblkid to get the UUID for the system disk

That one is really ugly, since it looks like there is no universal, portable way to uniquely identify a system...

**Comment by markpizz** [comment]
Hmmmm... 

1 & 2 are reasonable answers.

3 is more problematic.  Presuming that we've got a system UUID (as you suggest), and a possible NIC  instance Unique signature  Rather than the config file, the path to the MAC storage file is better since I've already got config files which drive different simulator instances simultaneously with differences be supplied by command line parameters.  In any case, even with some system and NIC random data, the range for random NIC generation is certainly not a 48bit MAC address since most MAC addresses ranges are allocated to various vendors or other purposes and as such generating random MAC values in those ranges would be at least unfriendly, and certainly violate numerous standards.  Alternatively a block could be registered to this project and addresses could be allocated randomly from within that block.  This is feasible, BUT block registrations cost from $680 to ~$2700 depending on the size of the block.

You suggest that ARM based small computers assign random MACs to their Ethernet devices.  Can you point at some place where this is described in detail...

**Comment by Rhialto** [comment]
When reading the suggesion of a file containing the MAC address, I
immediately thought that that could also be used to specify a pattern to
use as a base for a MAC address. For instance, if the file contained
01:02:03:xx:xx:xx, then the first 3 bytes would be fixed as given and
only the latter 3 bytes would be generated.

Possibly this could be extended to allow more fine-grained bits to be
selected. This could for instance be used to set the U/L bit to Local,
the unicast/multicast bit to unicast, and leave all other bits free for
random assignment.

Alternatively this pattern could be a parameter of the same command that
refers to the file.

**Comment by jguillaumes** [comment]
> 3 is more problematic. Presuming that we've got a system UUID (as you suggest), and a possible NIC instance Unique signature Rather than the config file, the path to the MAC storage file is better since I've already got config files which drive different simulator instances simultaneously with differences be supplied by command line parameters. In any case, even with some system and NIC random data, the range for random NIC generation is certainly not a 48bit MAC address since most MAC addresses ranges are allocated to various vendors or other purposes and as such generating random MAC values in those ranges would be at least unfriendly, and certainly violate numerous standards. Alternatively a block could be registered to this project and addresses could be allocated randomly from within that block. This is feasible, BUT block registrations cost from $680 to ~$2700 depending on the size of the block.

I see. In practice, I “borrow” addresses from the DEC block (I guess that now pertains to HP). As Rhialto suggested, perhaps a PART of the address should be left to the user, and the rest generated semi-randomly. It is not a 100% solid solution, but it probably would be “fine enough” for most of simh use cases.

> You suggest that ARM based small computers assign random MACs to their Ethernet devices. Can you point at some place where this is described in detail…
> 
I own a Cubietruck and an Odroid-C1, and both use “radom” MAC addresses for their NICs. I’ve tried to find some in-deep reference, but the best I’ve found is this:

http://linux-sunxi.org/Ethernet

There are plenty of recipes around to hack the u-Boot configuration to get fixed MAC addresses, I’m fine with random MACs (the biggest issue is some instability until the ARP caches have been updated with the new MAC, and I hardly reboot my boxes…), so I have not gone deep with the issue.

**Comment by markpizz** [comment]
How about the following syntax:

    sim> SET XQ MAC=aa:bb:cc:dd:ee:ff{/bits}{>filespec}

where:

1. all of the aa:bb:cc:dd:ee:ff values must be hex digits
2. bits is the number of bits which are to be taken from the supplied MAC aa:bb:cc:dd:ee:ff with legal values from 16 to 48 and a default of 48 bits.
3. filespec specifies a file which contains the MAC address to be used and if it doesn't exist an appropriate generated address will be stored in this file and a subsequent SET MAC invocation with the same file used the file contained value rather than generating a new MAC

**Comment by markpizz** [comment]
This implies that merely specifying a bits value between 16 and 47 will cause some of the MAC address bits to be generated without regard to whether the generated value is stored for subsequent reuse.

Uniqueness information during the generation process will leverage:
- system unique info (UUID as previously described when available)
- simulator name
- simulator device
- current working directory
- file path of retention file.
- number of bits being generated.
- the specified base MAC address

**Comment by jguillaumes** [comment]
Let me rephrase your explanation to check if I understand what are you proposing.

1) SET XQ MAC=08:00:2B:AA:BB:CC  => sets the MAC address to the specified value, like now
2) SET XQ MAC=08:00:2B:AA:BB:CC/24 => Generates a new address in the "DEC" block
3) SET XQ MAC=08:00:2B:AA:BB:CC/24(mymac.dat) =>  Looks if mymac contains a valid MAC address. If so, it will use it, Else proceed as in (2), and store the generated MAC in mymac.dat
4) SET XQ MAC=08:00:2B:AA:BB:CC(mymac.dat) => This does not make sense, since we are specifying the whole address

08:00:2B:AA:BB:CC/48 is identical to 08:00:2B:AA:BB:CC

I'm not sure what your comment "it implies that..." means. Does it mean that specifying /bits will override the stored address in myfile.dat?

On the other hand, perhaps it would be easier to allow the user to specify a _partial_ address, like in:

SET XQ MAC=08:00:2B:

And assume the /24 specification in this case. I'm not sure if this would complicate the parsing mechanism in excess (the /bits solution is fine and it looks a lot like a CIDR specification).

**Comment by markpizz** [comment]
The file, if it exists will override the rest of the information.  If i doesn't exist, the generated address info will be stored for potential future use.  This will facilitate retaining the MAC details when a VM environment is moved from one host to another or from one place within a host to another.

The MAC generation logic will produce the same MAC values as long as the list of "Uniqueness" information hasn't change from one simulator invocation to the next.  And these will most likely be different for multiple simulators running on the same or different hosts.  As such, there may not be a need to save generated info into a file.

    2) SET XQ MAC=08:00:2B:AA:BB:CC/24 => Generates a new address in 
       the "DEC" block

I'm actually thinking of making the equivalent of this command happen for each of XQ, XQB, XU and XUB upon simulator startup.  

    3) SET XQ MAC=08:00:2B:AA:BB:CC/24(mymac.dat) => Looks if mymac 
       contains a valid MAC address. If so, it will use it, Else proceed as in (2), 
       and store the generated MAC in mymac.dat

Actually, if the specified file exists but contains garbage then the command will fail.

    4) SET XQ MAC=08:00:2B:AA:BB:CC(mymac.dat) => This does not make 
       sense, since we are specifying the whole address

True, but no harm is done if the configuration is written to the specified file...

**Comment by jguillaumes** [comment]
> El 24 maig 2016, a les 1:55, Mark Pizzolato <notifications@github.com> va escriure:
> 
> (…)
> 
> True, but no harm is done if the configuration is written to the specified file...
> 
> 

OK to all your comments. All the usability and use cases are covered.

**Comment by markpizz** [comment]
The code is there, give it a try.

**Comment by jguillaumes** [comment]
I've done some testing, and it seems to work very well. Outstanding work, as always. Just for the record, the help text has not been updated, but I guess you'll wait until the coding is nailed (and I think it is!).

When you consider it done I'll close the issue.

**Comment by markpizz** [comment]
I think the current help is indeed correct.  Let me know if you have other suggested details.  

I updated things earler, and now issue somewhat explanatory messages when invalid MAC address input is given.

**Comment by jguillaumes** [comment]
Uhh... It seems the MAC filespec is being capitalized (ie, I've specified pdp11-xq.mac and I get PDP11-XQ.MAC). Is that a desired behaviour?

**Comment by markpizz** [comment]
Well, you should use a computer that has a file system which is case=insensitive.  :-)

Give the latest code a try.

**Comment by markpizz** [comment]
OK to close this?

**Comment by jguillaumes** [comment]
Mission acomplished. Thanks!
